File: 606.根据二叉树创建字符串.py
#
# @lc app=leetcode.cn id=606 lang=python3
#
# [606] 根据二叉树创建字符串
#
# https://leetcode-cn.com/problems/construct-string-from-binary-tree/description/
#
# algorithms
# Easy (54.44%)
# Likes:    143
# Dislikes: 0
# Total Accepted:    15.7K
# Total Submissions: 28.7K
# Testcase Example:  '[1,2,3,4]'
#
# 你需要采用前序遍历的方式，将一个二叉树转换成一个由括号和整数组成的字符串。
#
# 空节点则用一对空括号 "()" 表示。而且你需要省略所有不影响字符串与原始二叉树之间的一对一映射关系的空括号对。
#
# 示例 1:
#
#
# 输入: 二叉树: [1,2,3,4]
# ⁠      1
# ⁠    /   \
# ⁠   2     3
# ⁠  /
# ⁠ 4
#
# 输出: "1(2(4))(3)"
#
# 解释: 原本将是“1(2(4)())(3())”，
# 在你省略所有不必要的空括号对之后，
# 它将是“1(2(4))(3)”。
#
#
# 示例 2:
#
#
# 输入: 二叉树: [1,2,3,null,4]
# ⁠      1
# ⁠    /   \
# ⁠   2     3
# ⁠    \
# ⁠     4
#
# 输出: "1(2()(4))(3)"
#
# 解释: 和第一个示例相似，
# 除了我们不能省略第一个对括号来中断输入和输出之间的一对一映射关系。
#
#
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def tree2str(self, t: TreeNode) -> str:
        # 迭代
        if t == None:
            return ""
        stack = [t]
        output = []
        while stack:
            item = stack.pop()
            if item in [')', '(', '()']:
                output.append(item)
            else:
                node = item
                output.append(str(node.val))
                if node.right:
                    stack.extend((')',node.right,'('))
                    if not node.left:
                        stack.append('()')
                if node.left:
                    stack.extend((')',node.left,'('))
        return ''.join(output)

        # 递归
        def build(node):
            if not node:
                return ''
            l = build(node.left)
            r = build(node.right)
            ans = f'{node.val}'
            ans += f'({l})' if l or r else ''
            ans += f'({r})' if r else ''
            return ans
        return build(t)


# @lc code=end

File: 606.根据二叉树创建字符串.go
/*
 * @lc app=leetcode.cn id=606 lang=golang
 *
 * [606] 根据二叉树创建字符串
 *
 * https://leetcode.cn/problems/construct-string-from-binary-tree/description/
 *
 * algorithms
 * Easy (62.06%)
 * Likes:    387
 * Dislikes: 0
 * Total Accepted:    76.3K
 * Total Submissions: 123.1K
 * Testcase Example:  '[1,2,3,4]'
 *
 * 给你二叉树的根节点 root ，请你采用前序遍历的方式，将二叉树转化为一个由括号和整数组成的字符串，返回构造出的字符串。
 *
 * 空节点使用一对空括号对 "()" 表示，转化后需要省略所有不影响字符串与原始二叉树之间的一对一映射关系的空括号对。
 *
 *
 *
 *
 *
 * 示例 1：
 *
 *
 * 输入：root = [1,2,3,4]
 * 输出："1(2(4))(3)"
 * 解释：初步转化后得到 "1(2(4)())(3()())" ，但省略所有不必要的空括号对后，字符串应该是"1(2(4))(3)" 。
 *
 *
 * 示例 2：
 *
 *
 * 输入：root = [1,2,3,null,4]
 * 输出："1(2()(4))(3)"
 * 解释：和第一个示例类似，但是无法省略第一个空括号对，否则会破坏输入与输出一一映射的关系。
 *
 *
 *
 * 提示：
 *
 *
 * 树中节点的数目范围是 [1, 10^4]
 * -1000 <= Node.val <= 1000
 *
 *
 *
 *
 */
package main

import "strings"
import "strconv"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func tree2str(root *TreeNode) string {
	// ans := []string{}
	ans := strings.Builder{}
	var dfs func(node *TreeNode)
	dfs = func(node *TreeNode) {
		if node == nil {
			return
		}
		ans.WriteString(strconv.Itoa(node.Val))
		// 如何组织条件语句?
		// 1. 按条件组织行为
		// 2. 按行为组织条件
		if node.Left != nil || node.Right != nil {
			ans.WriteString("(")
			dfs(node.Left)
			ans.WriteString(")")
		}
		if node.Right != nil {
			ans.WriteString("(")
			dfs(node.Right)
			ans.WriteString(")")
		}
	}
	dfs(root)
	return ans.String()
}

// @lc code=end
